### PR TITLE
fix: no-delay on first re-connect attempt

### PR DESCRIPTION
**Pull request by zmstone**

### CODE REVIEW

File: src/wolff_producer.erl [hunk 1]
```diff
@@ -456,11 +466,14 @@ mark_connection_down(#{topic := Topic,
   case is_idle(St) of
     true ->
       %% nothing to send, will trigger reconnect later
+      log_info(Topic, Partition, "lost connection to partition leader while idling, "
```
Comment (qzhuyan) [inline src/wolff_producer.erl:469]: Will this trigger lots of logging? Could we use counters instead?

**Review by qzhuyan — APPROVED** [review]

File: src/wolff_producer.erl (hunk 1 above)
Comment (zmstone) [reply src/wolff_producer.erl:469]: This is to make the log symmetric.
There is a log for retry, but not one down.
When it comes to number of logs:
It’s not going to be more than the other exiting reconnect attempt log.

File: src/wolff_producer.erl [hunk 2]
```diff
@@ -478,13 +491,20 @@ ensure_delayed_reconnect(#{config := #{reconnect_delay_ms := Delay0},
...
   Attempts > 0 andalso Attempts rem 10 =:= 0 andalso
     log_error(Topic, Partition, "still disconnected after ~p reconnect attempts", [Attempts]),
-  %% add an extra random delay to avoid all partition workers
-  %% try to reconnect around the same moment
-  Delay = Delay0 + rand:uniform(1000),
+  Delay =
+    case DelayStrategy of
+      no_delay_for_first_attempt when Attempts =:= 0 ->
+        %% this is the first attempt after disconnected, no delay
+        0;
```
Comment (turtleDeng) [inline src/wolff_producer.erl:502]: If the delay is 0, do I still need to print this log?

```
log_info(Topic, Partition, "will try to rediscover leader connection after ~p ms delay", [Delay]),
```

Comment (zmstone) [reply src/wolff_producer.erl:502]: yes. there will be a "lost connection" log, we should log the attempt to make it symmetrical

**Review by k32 — APPROVED** [review]

**Pull request merged**